tensorbay/cli/log.py

```diff
@@ -24,32 +27,105 @@
+) -> Iterator[str]:
+    # Already printed commits set.
+    printed_commits: Set[str] = set()
+    # The amount of branches.
+    branch_amount: int = len(valid_branch_name)
+    # All commits paging list.
+    all_commits: List[PagingList[Commit]] = [
+        dataset_client.list_commits(branch_name) for branch_name in valid_branch_name
+    ]
+    # Print pointer of branches.
+    branch_pointer: List[int] = [0] * branch_amount
+    # Branch name of leaves.
+    leaves_to_branches: DefaultDict[str, Set[str]] = defaultdict(set)
+    # Printing commit list.
+    current_commits: List[Commit] = list()
+    for index, commits in enumerate(all_commits):
+        leaves_to_branches[commits[0].commit_id].add(valid_branch_name[index])
+        current_commits.append(commits[0])
+
+    while True:
+        commit, index = _get_newest_commit(current_commits)
+        commit_id = commit.commit_id
+        if commit_id in printed_commits:
+            # Merge branches.
+            del all_commits[index]
+            del current_commits[index]
+            del branch_pointer[index]
+            continue
+        # Reach root commit.
+        print_branch_name = None
+        if commit_id in leaves_to_branches:
+            print_branch_name = ", ".join(leaves_to_branches[commit_id])
+        if commit.parent_commit_id == ROOT_COMMIT_ID:
+            printed_commits.add(commit_id)
+            yield print_func(commit, print_branch_name)
+            break
+        branch_pointer[index] += 1
+        current_commits[index] = all_commits[index][branch_pointer[index]]
+        printed_commits.add(commit_id)
+        yield print_func(commit, print_branch_name)
+
+
+def _implement_log(  # pylint: disable=too-many-locals
     obj: Dict[str, str],
     tbrn: str,
     max_count: Optional[int],
     oneline: bool,
+    is_all: bool,
 ) -> None:
     gas = get_gas(**obj)
     info = TBRN(tbrn=tbrn)
     dataset_client = get_dataset_client(gas, info)
 
-    commits = dataset_client.list_commits(info.revision)[:max_count]
+    if is_all:
+        branches = dataset_client.list_branches()
+        valid_branch_name: List[str] = [branch.name for branch in branches]
+    else:
+        branch_name = dataset_client.status.branch_name
+        if not branch_name:
+            error("Aborting gas log due to empty branch name")
+        valid_branch_name = [branch_name]
     template = _get_oneline_log if oneline else _get_full_log
-    click.echo_via_pager(template(commit) for commit in commits)
+    print_commits = _print_all_commits(dataset_client, valid_branch_name, template)
+    click.echo_via_pager(islice(print_commits, max_count))
 
 
-def _get_oneline_log(commit: Commit) -> str:
-    return f"{shorten(commit.commit_id)} {commit.title}\n"
+def _get_oneline_log(commit: Commit, branch_name: Optional[str]) -> str:
+    commit_id = shorten(commit.commit_id)
+    if branch_name:
+        commit_id += f"({branch_name})"
```

Comment:
just use fstring, `str.__iadd__` is highly not recommended

Reply:
OK~